### PR TITLE
feat: add timezone option to sync window

### LINKED ISSUE #3204: Add timezone option in Sync Window
# Summary

Add timezone to Sync Window to handle daylight savings.

# Motivation

For distributed team working across multiple geographies, if two sync windows of duration 2 hrs need to be added, one for IST (Kolkata, India) and another for EST timezone, it needs to be converted to UTC. Let's assume, the sync window starts at 10AM local time for both the timezones, the sync window schedule can be converted to UTC as below.
30 4 * * *
0 15 * * *

When daylight savings starts, the first sync window is still good since daylight savings is not implemented in IST. However, the second sync window will then become active from 11 AM EST time, which is an hour late.

As of now, someone has to reconfigure the sync windows right before daylight savings kicks in.

# Proposal

Add an option in UI to take timezone as input.
The backend server should convert the sync window with respective timezone to UTC to analyze if it is active or not.
The backend server should have the intelligence to consider daylight savings while converting it to UTC.
For backward compatibility, all existing sync windows timezone should be considered as UTC

### PR DESCRIPTION
**Pull request by ishitasequeira**
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 

Closes #3204

Signed off by [Ishita Sequeira](@ishitasequeira )

### CODE REVIEW

**Review by reginaelyse — APPROVED** [review]
got it to work, lgtm :)
 
 timezones1
 
timezones2

**Review by jannfis — CHANGES_REQUESTED** [review]
Thanks @ishitasequeira, LGTM with a few minor things (see below).

Also, I'm in line with @reginapizza's comment about the wording. According to https://www.merriam-webster.com/dictionary/time%20zone, these are two words. Can you please change the human readable occurances to "time zone" (in the UI and the CLI) instead of using "TimeZone"? 

To be consistent with other CLI commands and arguments, can you please use `--time-zone` instead of `--timeZone` as parameter name in the CLI?

**Comment by jannfis** [comment]
Thanks for the update, @ishitasequeira. However, I still think "Time Zone" (capitalized) isn't correct and feels a little weird. 

Can you please update strings to:

* `Time zone` at the beginning of a sentence (only Time should be capitalized there)
* `time zone` (all lowercase) otherwise

Keeping the naming of the variables in camelCase (e.g. `timeZone` for private vars and `TimeZone` for public ones) is fine, since it's two words.

This may feel like a nit, but imho improves readability and user experience as well.

**Comment by ishitasequeira** [comment]
@jannfis It makes sense to update the "Time zone" convention and the variable name convention. I have made the required changes. PTAL.

Screen Shot 2021-11-01 at 4 06 35 PM
Screen Shot 2021-11-01 at 5 03 14 PM

File: assets/swagger.json [hunk 1]
```diff
@@ -4899,6 +4899,10 @@
             "$ref": "#/definitions/v1alpha1HelmParameter"
           }
         },
+        "passCredentials": {
+          "type": "boolean",
+          "title": "PassCredentials pass credentials to all domains (Helm's --pass-credentials)"
+        },
```
Comment (jannfis) [inline assets/swagger.json:4905]: Where does this come from?

File: pkg/apis/application/v1alpha1/types.go [hunk 2]
```diff
@@ -1952,12 +1977,24 @@ func (w *SyncWindow) Update(s string, d string, a []string, n []string, c []stri
...
 
 // Validate checks whether a sync window has valid configuration. The error returned indicates any problems that has been found.
 func (w *SyncWindow) Validate() error {
+
+	// Default timeZone to UTC if timeZone is not specified
+	if w.TimeZone == "" {
+		w.TimeZone = "UTC"
+	}
+	if _, err := time.LoadLocation(w.TimeZone); err != nil {
+		return fmt.Errorf("timeZone '%s' is invalid", w.TimeZone)
```
Comment (jannfis) [inline pkg/apis/application/v1alpha1/types.go:1995]: You seem to have missed this one

File: assets/swagger.json (hunk 1 above)
Comment (ishitasequeira) [reply assets/swagger.json]: Thanks for catching this. Not sure how this got added. I have removed this part.

**Review by jannfis — APPROVED** [review]
LGTM, thank you @ishitasequeira !

**Pull request merged**